students/stefan_lund/Lesson_5/Lesson_5_mailroom/mailroom_5_functions.py

```diff
@@ -0,0 +1,323 @@
+    p = "+"
+    s = "-"
+    lne = s * (name_len)
+    e = p + (s * (amount_len))
+    lne += 3 * e
+    print(lne)
+
+    for name, donations in data.items():
+        number_of = 0
+        donation = 0
+        for date in donations:
+            donation += donations[date]
+            number_of += 1
+
+            total = float(donation)
+            number_of = int(number_of)
+            average = total / number_of
+
+        temp_form = "{:<{nl}} ${:>{al}.2f}  {:>{al}} ${:>{al}.2f}"
+        line = temp_form.format(name,
+                                total,
+                                number_of,
+                                average,
+                                nl=name_len,
+                                al=amount_len - 1)
+        print(line)
+
+    return "start"
+
+
+def send_letters_to_everyone():
+    """
+        sends one type of letter chosen from a template of letters to
+            all names in the data dictionary
+        each letter is stored as a txt file in cwd folder "letter_to_everyone"
+    """
+    data = mailroom_5_read_write_data.data
+    letter_template = letter_templates("2")  # needs date, name, and total amount
+    today = todays_date()
+    year, month, day = today[0], today[1], today[2]
+
+    letters = []
+    # create letters
+    for name, donations in sorted(data.items()):
+        donation = 0
+        for date in donations:
+            donation += donations[date]
+
+        total = float(donation)
+
+        letter = letter_template.format(name=name,
+                                        amount=total,
+                                        date=month+"/"+day+"/"+year)
+        donor = "_".join(name.split())
+        letter_file = donor + "__" + year + "_" + month + "_" + day
+        letters.append((letter, letter_file))
+
+
+    # print letter to screen and store letter
+
+    # create the directory to store the letters if it doesn't exist
+    pth = os.path.join(os.getcwd(), "letter_to_everyone")
+    if not os.path.exists(pth):
+        os.makedirs(pth)
+
+    for letter, letterfile in letters:
+        print("\n", letter)
+        destination = os.path.join(pth, letterfile)
+
+        try:
+            with open(destination, 'w') as outfile:
+                for line in letter:
+                    outfile.write(line)
+        except FileNotFoundError:
+            print("Could not find the file")
+
+    print("stored letters in ", pth)
+
+    return "start"
+
+
+# "thank you" menu functions
+
+def main_menu():
+    """
+    sole function is to switch to the "start" menue
+    """
+    return "start"
+
+def list_of_names():
+    """
+        prints out a list of names currently in the "data" dictionary
+    """
+    data = mailroom_5_read_write_data.data
+
+    # determine width of fields to print in
+    name_list = data.keys()
+    name_length = []
+    # list comprehension
+    [name_length.append(len(name)) for name in name_list]

```

Comment:
this is incorrect usage of comprehension; see below for correct usage:
```
name_lengths = [len(name) for name in data]
```

Reply:
# name_list = data.keys()
# name_length = []
# (wrong usage of )list comprehension
# [name_length.append(len(name)) for name in name_list]
# name_len = max(name_length) + 3
becomes
name_len = max([len(name) for name in data]) + 3

Thank you for showing me. 
I knew something was off but could not get it right. Much nicer and not only the comprehension but the use of data-dict as well.